### PR TITLE
support delta along with ANSI support

### PR DESCRIPTION
**Pull request by ulwlu**
https://github.com/jonas/tig/issues/542#issuecomment-899094741

- All test passes.
- I haven't seen segmentation faults with newest versions of MacOS Big Sur, iTerm2, and delta. ncurses version is 6.2.
- It looks like no performance issues.
- I've only tested it with Terminal.app and iTerm on MacOS.

ss 0003-08-17 at 19 11 58
ss 0003-08-17 at 19 12 36

### how to use delta in tig

It requires ncurse 6.1 or higher.

```bash
# if you don't have ncurses
curl -O ftp://ftp.gnu.org/gnu/ncurses/ncurses-6.2.tar.gz
tar -xzvf ncurses-6.2.tar.gz
cd ncurses-6.2
./configure --prefix=${where you like} --enable-pc-files --with-pkg-config-libdir=/usr/local/bin/pkgconfig --enable-sigwinch --enable-symlinks --enable-widec --with-shared --with-gpm=no --without-ada --enable-ext-colors # i set the path as /usr/local which is default
sudo make
sudo make install

cd ${path to tig of the branch of above PR}
make configure
./configure LDFLAGS=-L${where you download ncurses}/lib CPPFLAGS=-I{where you installed ncurses}/include  # i set the path as /usr/local/opt/ncurses which is installed by ncurses
make

echo "set diff-highlight = \"delta\"" >> ~/.tigrc

./src/tig
```

### CODE REVIEW

**Comment by ulwlu** [comment]
This now works.

ss 0003-08-17 at 19 11 58

ss 0003-08-17 at 19 12 36

**Comment by hasufell** [comment]
Great stuff.

How is the highlighting expected to work? I found it a bit surprising:

Peek 2021-08-17 17-36

**Comment by hasufell** [comment]
I also seem to get some line skew with very long lines:

Peek 2021-08-17 17-51

**Comment by ulwlu** [comment]
Hi hasufell, thank you for catching. I'll fix those.

**Comment by ronjouch** [comment]
@ulwlu hi, here to respond to your call for tests at https://github.com/jonas/tig/issues/542#issuecomment-900179550 .

Your branch compiles and works for me (Arch Linux / [Kitty](https://sw.kovidgoyal.net/kitty/) 0.23.1, ncurses 6.2, delta 0.8.3) 🙂.

In addition to the weird line highlight already reported, I was surprised to see delta-within-tig uses a gray background. I'm not sure why, as pure delta outside of tig uses a black background. My `delta` config is empty, and my `~/.tigrc` is empty except for the single line `set diff-highlight = "delta"`.

Screencast below (pure delta at the beginning from 0:00 to 0:07, delta-in-tig afterwards):

https://user-images.githubusercontent.com/522085/129816178-73f18a78-b889-4aae-ab09-73d91c7a605a.mp4

**Comment by ulwlu** [comment]
@ronjouch Hi, thank you for catching and details. I'm still not sure if this is caused by coloring problem, but I'm thinking delta gave some background color info with '--trur-color ＝never' that is not close to real true-color.

I'll check it anyway. Thank you very much

**Comment by ronjouch** [comment]
> Thank you very much

Much obliged; thank you for the work on this!

> I'm still not sure if this is caused by coloring problem, but I'm thinking delta gave some background color info with '--trur-color ＝never' that is not close to real true-color.

@ulwlu I did a bit more monkeying around with kitty color configuration. The problem doesn't appear by default. But with my Kitty color config (pasted below), it does.

```
#: The color table {{{

#: The 256 terminal colors. There are 8 basic colors, each color has a
#: dull and bright version, for the first 16 colors. You can set the
#: remaining 240 colors as color16 to color255.

color0 #555753
color8 #777777
#: black

color1 #ef2929
color9 #cc0000
#: red

color2  #8ae234
color10 #4e9a06
#: green

color3  #fce94f
color11 #c4a000
#: yellow

color4  #729fcf
color12 #3465a4
#: blue

color5  #ad7fa8
color13 #75507b
#: magenta

color6  #34e2e2
color14 #06989a
#: cyan

color7  #eeeeec
color15 #d3d7cf
#: white
```

File: .vscode/settings.json
Comment (pablospe) [inline .vscode/settings.json:7]: I believe this file shouldn't be part of the PR

**Comment by pablospe** [comment]
Another details. Have you tried **side-by-side** mode? It seems to work but only until 80 chars. So, I was thinking to use `delta --width 250`, but this doesn't seem to work: `set diff-highlight = "delta --width 250"`. How do you pass parameters?

I've also observed a weird highlighting, gray background. I am using konsole.

**Comment by pablospe** [comment]
Also noticed that when there are many files and you press enter to the filename (in the commit header), it doesn't jump to beginning of the file in the diff. Instead I received this error: 
```
[diff] Press '<Enter>' to jump to file diff - line 11 of 88                                                                                                                                                                                 
Failed to find file diff
```

**Comment by ulwlu** [comment]
@hasufell @ronjouch 

Hi. I've successfully fixed those, but I've come up with a better design. So I've decided to refactor it, which should take about a week.

@pablospe 

Tig can't accept formatter options. I can fix that but it will be on another PR since it has some impact on another function. Currently could you set config locally?

> I believe this file shouldn't be part of the PR

I knew what you want to say, but it would be a pain to develop if I separate PR for the vscode settings, so I'm doing this. If jonas tells about it, I'll separate it, but for now I won't.

> it doesn't jump to beginning of the file in the diff

Thank you for catching. I'll fix this.

**Comment by pablospe** [comment]
> Tig can't accept formatter options. I can fix that but it will be on another PR since it has some impact on another function. Currently could you set config locally?
> 
> >

Apologies, I didn't understand. What do you mean by set config locally? Having an alias or something like that to avoid options?

**Comment by ulwlu** [comment]
@pablospe 

Sorry for the lack of explanation, please do the following at this time.

```bash
git config delta.width 250
```

From your explanation, I think you don't want to set it to global. However, tig does not accept options now, so please set locally like this inside repositories that you use tig. Currently tig searches only executable binary, so alias doesn't work either. I will probably fix it to accept options if jonas agreed.

Also, in the original issue I asked what to do with the side-by-side option inside tig. If you have time, please feel free to post something like "allow tigrc to accept options and allow users to set width freely", so that someone may post an another idea.

https://github.com/jonas/tig/issues/542#issuecomment-899094741

**Comment by pablospe** [comment]
Also another problem I found is that in navigation mode for delta doesn't work:
`n and N keybindings to move between files in large diffs, and between diffs in log -p views (--navigate)`. See in delta documentation ([here](https://github.com/dandavison/delta#deltas-main-features-are))

I am just reporting the problems I found. Thanks a lot for the fantastic work, @ulwlu!

**Comment by krobelus** [comment]
I haven't checked the implementation details but this looks a bit weird on light terminals like gnome-terminal or xterm because it draws dark backgrounds.

**Comment by ulwlu** [comment]
@krobelus Hi. Yes that occured because of here.

https://github.com/jonas/tig/pull/1140/files#diff-3cd1efc51226a700b579928f3ee8477ed1a05715a60b2d6e0b1fc54e63969581R50


I noticed that and fixed it into default_bg. Thank you for catching.

P.S.
Currently I'm fixing a bit of functions which pablospe mentioned.

**Comment by pablospe** [comment]
> Currently I'm fixing a bit of functions which pablospe mentioned and more else.

Maybe some of the changes can be separated in other PRs and add basic `delta` support on master? Even if it is not perfect, it would be good to have the delta support.

**Comment by ulwlu** [comment]
@pablospe sorry to be late. I was busy with my work and time flies so fast. That sounds great, I wanna add at least one more feature then.

- [x] handle delta output and draw correct color
- [x] draw correct current cursor
- [x] calculate the correct line length ( so that the line won't be overwrapped)
- [ ] handle correct pager <- i want to add at least this one

These can be called basic feature, and we can add some features such as jumping to the file hunks.

I probably can add in this year

**Comment by ronjouch** [comment]
> @ronjouch I don’t have kitty env but I think I have fixed your case.

@ulwlu thanks, fix confirmed! I'll be running 629a12e and will report new problems here.

**Review by pablospe — APPROVED** [review]

**Comment by pablospe** [comment]
Looks good to me! Thanks for the great job, @ulwlu! 
@jonas Who can approve this PR to be merged?

**Comment by ulwlu** [comment]
thank you pablospe for your review!

probably @koutcher would you review this PR?

**Comment by mbsulliv** [comment]
I've compiled 5ae769b and am having a segmentation fault on Linux.  I compiled with "make all-debug" and then ran tig via gdb and the crash is happening in draw_ansi() on line 151: the `color_method_mark` variable is null.

The crash seems related to lines like the following in my ~/.gitconfig:
`minus-emph-style = bold red 52`

If a line like that is present (i.e. a line with a background color), then the crash happens pretty reliably.  Deleting the background color (i.e. the 52) from that line makes the crash go away.  Note that even if the crash goes away, I still see some erratic behavior where delta sometimes doesn't display the diff correctly.

**Comment by ulwlu** [comment]
@mbsulliv 
thank you for catching, i confirmed the bug.
It happens with the following string example, 

```
^[[34m^[[38;5;88m 28 ^[[34m⋮^[[38;5;28m    ^[[34m│^[[31mvoid draw_ansi(struct view *view, int *ansi_num, char **ansi_ptrs);^[[0m
```
.... because i should add the null check before the exact line where you pointed out `color_method_mark`.

this should be covered, i appreciate your catching. I'll fix and add a test as well. (until next Wednesday)

File: src/ansi.c
Comment (mbsulliv) [inline src/ansi.c:211]: I believe `save_ptr` needs to be a `char**` here because `strtok_r` modifies the value of the pointer.  As it is currently written, when the value of `save_ptr` gets modified, the changed value isn't passed back into the calling function.

I discovered this problem when it tripped over the following ANSI code: `\033[38;5;8m`.  The "38" triggered us to enter this function, the "5" was handled here, but upon returning, the "5" was parsed again and it caused all my code to start blinking.  It was pretty funny actually ;)

I tested this change on my side and things started behaving correctly.

File: src/ansi.c
Comment (mbsulliv) [inline src/ansi.c]: Is it intentional to delete the "47" code?

**Review by mbsulliv — CHANGES_REQUESTED** [review]

File: src/ansi.c
Comment (ulwlu) [reply src/ansi.c]: You are correct, sorry for my mistake. Fixed here [5093bb1](https://github.com/jonas/tig/pull/1140/commits/5093bb1f89ec4309956fd4101e46aa64cce485a8).

This could be an excuse, but the reason why I didn't notice is that iTerm and VSCode-Terminal doesn't support BLINK.... And Macbook-builtin Terminal does support. Now I confirmed the bug and fixed. 
https://unix.stackexchange.com/questions/151673/attrona-blink-using-ncurses-does-nothing 

Thank you very much for catching again.

File: src/ansi.c
Comment (ulwlu) [reply src/ansi.c]: Sorry, it's just my mistake. 
Thanks for catching, Fixed here b86f198 .

**Comment by fengyichui** [comment]
Hi,
In diff view, the file jump feature does not work properly and show error message: "Failed to find file diff"

**Comment by ulwlu** [comment]
@fengyichui 
please see the comment above, this supports the basic feature. 

```
Maybe some of the changes can be separated in other PRs and add basic delta support on master? Even if it is not perfect, it would be good to have the delta support.
```

**Comment by fengyichui** [comment]
@ulwlu  Sorry, I don't quite understand, do you mean it's already supported or need other PRs ?

**Comment by ulwlu** [comment]
it needs other PR.

**Comment by fengyichui** [comment]
@ulwlu  Thanks for your answer.

**Comment by edubxb** [comment]
Looking forward to seeing this merged, awesome, thanks!

**Comment by crokobit** [comment]
Can we merge this?

**Comment by FdelMazo** [comment]
Hey @koutcher, any chance we can get this merged? I'd love this feature

**Comment by ulwlu** [comment]
Thank you @koutcher for reviewing.
Your personal time should always come first, so never mind.

I think your considerations are all reasonable and I want to get started on them right away (especially the 256x256 fixed array, which I thought same thing when I created it. I didn't see any performance degradation in memory or speed after creation, but I'll look again to see if there are any other workarounds).

Actually, I will start working for a new company next week and will be a bit busy, so I will start next weekend,
but in terms of below question

> all these if (strcmp(... without elses, is it on purpose?

Yes. I didn't put else where else is not necessary.

Thanks again for taking the time to review this.

**Comment by koutcher** [comment]
@ulwlu, please do. As I'm not a delta or any other fancy diff user, I'll create a feature branch to allow contributions until it is mature enough to be merged. Thanks again for your time.

**Comment by ulwlu** [comment]
@koutcher 
just a note to let you know, I've already sent you a patch in email 12 days ago. (from ooulwluoo@gmail.com)
you don't need to reply this if you've noticed it.

**Comment by ijoseph** [comment]
For Apple Silicon (M1)/ Homebrew, I had to

```
./configure  LDFLAGS="-L/opt/homebrew/opt/ncurses/lib" CPPFLAGS="-I/opt/homebrew/opt/ncurses/include"  PKG_CONFIG_PATH="/opt/homebrew/opt/ncurses/lib/pkgconfig"
```


based on 

```
╰─ brew info ncurses
==> ncurses: stable 6.3 (bottled) [keg-only]
Text-based UI library
https://invisible-island.net/ncurses/announce.html
/opt/homebrew/Cellar/ncurses/6.3 (3,968 files, 9.6MB)
  Poured from bottle on 2022-03-17 at 16:24:05
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/ncurses.rb
License: MIT
==> Dependencies
Build: pkg-config ✔
==> Caveats
ncurses is keg-only, which means it was not symlinked into /opt/homebrew,
because macOS already provides this software and installing another version in
parallel can cause all kinds of trouble.

If you need to have ncurses first in your PATH, run:
  echo 'export PATH="/opt/homebrew/opt/ncurses/bin:$PATH"' >> ~/.zshrc

For compilers to find ncurses you may need to set:
  export LDFLAGS="-L/opt/homebrew/opt/ncurses/lib"
  export CPPFLAGS="-I/opt/homebrew/opt/ncurses/include"

For pkg-config to find ncurses you may need to set:
  export PKG_CONFIG_PATH="/opt/homebrew/opt/ncurses/lib/pkgconfig"
```

**Comment by FdelMazo** [comment]
I'm not entirely sure. Does this PR helps towards adding `--color-moved` support? Issue #890

**Comment by vitornesello** [comment]
This MR is pretty cool. Does anyone know when it will be released?

**Comment by aberres** [comment]
Someone else just being here to ask if there are plans to get this in  😇

**Comment by pablospe** [comment]
In the meanwhile, I ended up using this shortcuts (in the .tigrc):

```
# Open delta with git show.
bind generic d >sh -c "DELTA_PAGER='less -RKc' git show %(commit)"

# https://github.com/jonas/tig/issues/542#issuecomment-1242039892
bind generic D @sh -c "\
  ( \
    tmux has-session -t \".{last}\" \
    && tmux respawn-pane -t \".{last}\" -k 'LESS= DELTA_PAGER=\"less -R\" git show %(commit)' \
  ) \
  || tmux split-window -l 50% 'LESS= DELTA_PAGER=\"less -R\" git show %(commit)'"
```

First shortcut `d` is a generic to see the diff in full-screen. If you are inside tmux, you could also try the `D` shortcut, it is more convenient since it opens a new pane and you can see the list of commits too, but usually I end up using `d` in general because it is full-screen and I see more. What it is missing is a way to auto-detect if you are inside tmux, and select the proper behavior, rather than having two shortcuts. 

Obviously this is PR is a better solution, but for now this is a workaround that works for me. I hope this PR get merge soon.

**Comment by pablospe** [comment]
Can this PR be re-open? @koutcher @jonas 
(the author account was deleted)

Alternative I could open a new PR from my github fork. I will wait some time before creating the PR myself, in case this PR could be re-open easily .

**Comment by koutcher** [comment]
@pablospe, GH does not allow to reopen a PR from a deleted account. The latest status from the original author is already accessible in the ansi-support branch: https://github.com/jonas/tig/tree/ansi-support.

**Comment by pablospe** [comment]
In that case, I will create a new PR pointing: https://github.com/jonas/tig/tree/ansi-support.

Few questions:
    * From where is the [Rework after review](https://github.com/jonas/tig/commit/d2b29cd2d424ccdd78ef080bd6b8e184847ce909) commit is coming from, it doesn't seem to be part of this PR.
    * What would be the next steps to get this PR in?

**Comment by adamency** [comment]
Hi @pablospe, I've seen your fork https://github.com/pablospe/tig.

Can it be used as is (i.e. simply replacing the `tig` executable) to be able to use `delta` in tig ?

**Comment by adamency** [comment]
I'm leaving this comment here for people who want to try delta in `tig` now. (The author) @uwlwu's instruction were not very precise.

## Installation Instructions

```
git clone https://github.com/jonas/tig/ 
cd tig
# Checkout branch of the project which contains the PR additions
git checkout ansi-support
make configure
# `ncurses` is almost always installable by your package manager, and almost always already installed on your system as it is needed for many TUI programs. The paths given here are the ones used by the package manager.
./configure LDFLAGS="-L/usr/lib" CPPFLAGS="-I/usr/include"
# This will install `tig` to ~/.local/bin, the recommended user path by the XDG specs. If needed, replace it by the path you want but leave out the `/bin/` part as this will be added by `make` already.
make prefix=$HOME/.local/
make install prefix=$HOME/.local/
# Configure `tig` to use `delta` in diffs
echo "set diff-highlight = \"delta\"" >> ~/.tigrc
```

This should work. However I have to note, some commits are displaying with the normal output instead of `delta`'s one. I don't know what's going wrong in theses cases.

**Comment by pablospe** [comment]
Let's move the discussion to:
**PR https://github.com/jonas/tig/pull/1298:** [Support delta along with ANSI support (continuation)](https://github.com/jonas/tig/pull/1298) 

**Notes**: 
1. Why? GH does not allow to reopen a PR from a deleted account.
2. @koutcher I didn't have permission to create a new PR from https://github.com/jonas/tig/tree/ansi-support, so I did it from my fork, hope this is ok with you. What are the next steps?

**Pull request closed without merging**
